swarm/src/lib.rs

```diff
@@ -415,19 +415,43 @@ where
 
         let peer_id = dial_opts.get_peer_id();
         let condition = dial_opts.peer_condition();
+        let dial_conditions = dial_opts.dial_conditions();
         let connection_id = dial_opts.connection_id();
 
-        let should_dial = match (condition, peer_id) {
-            (PeerCondition::Always, _) => true,
-            (PeerCondition::Disconnected, None) => true,
-            (PeerCondition::NotDialing, None) => true,
-            (PeerCondition::Disconnected, Some(peer_id)) => !self.pool.is_connected(peer_id),
-            (PeerCondition::NotDialing, Some(peer_id)) => !self.pool.is_dialing(peer_id),
+        let dial_failure = if let Some(peer_id) = peer_id {
+            match condition {
+                // Only use deprecated `peer_condition` when the `Option` was explicitly set.
+                Some(condition) => match condition {
+                    PeerCondition::Disconnected if !self.pool.is_connected(peer_id) => Some(
+                        DialError::DialPeerConditionFalse(PeerCondition::Disconnected),
+                    ),
+                    PeerCondition::NotDialing if !self.pool.is_dialing(peer_id) => {
+                        Some(DialError::DialPeerConditionFalse(PeerCondition::NotDialing))
+                    }
+                    _ => None,
+                },
+                None => {
+                    if dial_conditions.contains(DialConditions::Disconnected)
+                        && self.pool.is_connected(peer_id)
+                    {
+                        Some(DialError::DialPeerConditionFalse(
+                            PeerCondition::Disconnected,
+                        ))
+                    } else if dial_conditions.contains(DialConditions::NotDialing)
+                        && self.pool.is_dialing(peer_id)
+                    {
+                        Some(DialError::DialPeerConditionFalse(PeerCondition::NotDialing))
+                    } else {
+                        None
+                    }
+                }
+            }
+        // Conditions are not used without a `PeerId`.
+        } else {
+            None
         };
 
-        if !should_dial {
-            let e = DialError::DialPeerConditionFalse(condition);
-
+        if let Some(e) = dial_failure {
             self.behaviour
```

Comment:
I don't feel strongly about keeping this DRY. Can we split the deprecated code further from the current logic if we just duplicate the call to `on_swarm_event` and then call `return (e)`?